<h1>1 Data Acquisition and Preprocessing</h1>

<h3>1.1 Data Acquisition</h3>

<h4>Data resource:</h4>
<p>"Census Income Data Set": https://archive.ics.uci.edu/ml/datasets/Census+Income</p>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn

In [2]:
names = pd.read_csv("adult.names", sep="\t")

In [3]:
names

| This data was extracted from the census bureau database found at
0    | http://www.census.gov/ftp/pub/DES/www/welcom...                
1              | Donor: Ronny Kohavi and Barry Becker,                
2               |        Data Mining and Visualization                
3                           |        Silicon Graphics.                
4       |        e-mail: ronnyk@sgi.com for questions.                
..                                                 ...                
101                                 sex: Female, Male.                
102                          capital-gain: continuous.                
103                          capital-loss: continuous.                
104                        hours-per-week: continuous.                
105  native-country: United-States, Cambodia, Engla...                

[106 rows x 1 columns]

<h4>Observation: </h4>
<p>The adult.names file contains general information about the dataset, which cannot be used directly as columns names for the adult.data.</p>

In [4]:
data = pd.read_csv("adult.data")

In [5]:
data.head()

39          State-gov   77516   Bachelors   13        Never-married  \
0  50   Self-emp-not-inc   83311   Bachelors   13   Married-civ-spouse   
1  38            Private  215646     HS-grad    9             Divorced   
2  53            Private  234721        11th    7   Married-civ-spouse   
3  28            Private  338409   Bachelors   13   Married-civ-spouse   
4  37            Private  284582     Masters   14   Married-civ-spouse   

         Adm-clerical   Not-in-family   White     Male   2174   0   40  \
0     Exec-managerial         Husband   White     Male      0   0   13   
1   Handlers-cleaners   Not-in-family   White     Male      0   0   40   
2   Handlers-cleaners         Husband   Black     Male      0   0   40   
3      Prof-specialty            Wife   Black   Female      0   0   40   
4     Exec-managerial            Wife   White   Female      0   0   40   

    United-States   <=50K  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3            Cuba   <=50K  
4   United-States   <=50K

In [6]:
data.shape

(32560, 15)

<h4>Observation: </h4>
<p>The file "adult.data" has no header.</p>

In [7]:
data_test = pd.read_csv("adult.data")

In [8]:
data_test.head()

39          State-gov   77516   Bachelors   13        Never-married  \
0  50   Self-emp-not-inc   83311   Bachelors   13   Married-civ-spouse   
1  38            Private  215646     HS-grad    9             Divorced   
2  53            Private  234721        11th    7   Married-civ-spouse   
3  28            Private  338409   Bachelors   13   Married-civ-spouse   
4  37            Private  284582     Masters   14   Married-civ-spouse   

         Adm-clerical   Not-in-family   White     Male   2174   0   40  \
0     Exec-managerial         Husband   White     Male      0   0   13   
1   Handlers-cleaners   Not-in-family   White     Male      0   0   40   
2   Handlers-cleaners         Husband   Black     Male      0   0   40   
3      Prof-specialty            Wife   Black   Female      0   0   40   
4     Exec-managerial            Wife   White   Female      0   0   40   

    United-States   <=50K  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3            Cuba   <=50K  
4   United-States   <=50K

In [9]:
data_test.shape

(32560, 15)

<h4>Observation: </h4>
<p>1.The file "adult.test" has no header either.</p>
<p>2.Comparied with "adult.data", "adult.test" has less instances.</p>
<p>3.There is no concrete indication that the the whole dataset is consist of instances from "adult.data" and "adult.test". To ensure the validation, the "adult.data" file will be used for the dataset to work on in the project.</p>

In [10]:
#give header names to the chosen dataset
header_names = ["age","workclass","fnlwgt","education","education-num","marital-status","occupation","relationship","race","sex","capital-gain","capital-loss","hours-per-week","native-country","income"]
df = pd.read_csv("adult.data", names = header_names)
df.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

<h3>1.2 data checking</h3>
<p>Take a genenral look at the dataset.</p>

In [11]:
df.shape

(32561, 15)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


<h4>Observation: </h4>
<p>The dataset has 6 numerical variables and 9 categorial varibales.</p>

In [13]:
#checking if there is any missing value
df.isnull().any().sum() 

0

<h4>Observation: </h4>
<p>No missing value is found.</p>

In [14]:
#checking if there are duplicated attributes in the dataset
print(df.duplicated().sum()) # 24 duplicates are found

24


<h4>Observation: </h4>
<p>24 duplicates are found by checking duplicates.</p>
<p>Since no ID is used to identify indivduals uniquely. It is possible that 24 instances sharing the same values for all 15 attributes. For data integrity, the duplicates are kept in the dataset.</p>

In [15]:
#Checking categorical variables in rows-string
print(df.describe(include = [object]))

       workclass education       marital-status       occupation relationship  \
count      32561     32561                32561            32561        32561   
unique         9        16                    7               15            6   
top      Private   HS-grad   Married-civ-spouse   Prof-specialty      Husband   
freq       22696     10501                14976             4140        13193   

          race    sex  native-country  income  
count    32561  32561           32561   32561  
unique       5      2              42       2  
top      White   Male   United-States   <=50K  
freq     27816  21790           29170   24720  


In [16]:
#Checking numerial variables in rows-string
print(df.describe(include = [np.number])) 

                age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000  


<h3>1.3 data cleaning</h3>
<p>Checking extra whitespaces and impossible values.</p>

<h4>Observation: </h4>
<p>Based on the attribute information found in https://archive.ics.uci.edu/ml/datasets/Census+Income. Workclass has 8 types, and "?" is not among them. Thus "?" needs to be handled as missing values.</p>

In [17]:
df["workclass"].value_counts()

 Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 ?                    1836
 State-gov            1298
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: workclass, dtype: int64

In [18]:
# replacing "?" with the most frequent value "Private"
df["workclass"] = df["workclass"].replace("?", "Private")

In [19]:
#checking the progress
df["workclass"].value_counts()

 Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 ?                    1836
 State-gov            1298
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: workclass, dtype: int64

In [20]:
df["education"].value_counts()

 HS-grad         10501
 Some-college     7291
 Bachelors        5355
 Masters          1723
 Assoc-voc        1382
 11th             1175
 Assoc-acdm       1067
 10th              933
 7th-8th           646
 Prof-school       576
 9th               514
 12th              433
 Doctorate         413
 5th-6th           333
 1st-4th           168
 Preschool          51
Name: education, dtype: int64

In [21]:
df["marital-status"].value_counts()

 Married-civ-spouse       14976
 Never-married            10683
 Divorced                  4443
 Separated                 1025
 Widowed                    993
 Married-spouse-absent      418
 Married-AF-spouse           23
Name: marital-status, dtype: int64

<h4>Observation: </h4>
<p>The column "education" and "marital-status" match with the attribute information found in https://archive.ics.uci.edu/ml/datasets/Census+Income.</p>

In [22]:
df["occupation"].value_counts()

 Prof-specialty       4140
 Craft-repair         4099
 Exec-managerial      4066
 Adm-clerical         3770
 Sales                3650
 Other-service        3295
 Machine-op-inspct    2002
 ?                    1843
 Transport-moving     1597
 Handlers-cleaners    1370
 Farming-fishing       994
 Tech-support          928
 Protective-serv       649
 Priv-house-serv       149
 Armed-Forces            9
Name: occupation, dtype: int64

<h4>Observation: </h4>
<p>Based on the attribute information found in https://archive.ics.uci.edu/ml/datasets/Census+Income. "occupation" has 14 types, and "?" is not among them. Thus "?" needs to be handled as missing values.</p>

In [23]:
# replacing "?" with the most frequent value "Prof-specialty"
df["occupation"] = df["occupation"].str.replace("?", "Prof-specialty")

<ipython-input-23-83790df562fb>:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df["occupation"] = df["occupation"].str.replace("?", "Prof-specialty")


In [24]:
#checking the progress
df["occupation"].value_counts()

 Prof-specialty       5983
 Craft-repair         4099
 Exec-managerial      4066
 Adm-clerical         3770
 Sales                3650
 Other-service        3295
 Machine-op-inspct    2002
 Transport-moving     1597
 Handlers-cleaners    1370
 Farming-fishing       994
 Tech-support          928
 Protective-serv       649
 Priv-house-serv       149
 Armed-Forces            9
Name: occupation, dtype: int64

In [25]:
df["relationship"].value_counts()

 Husband           13193
 Not-in-family      8305
 Own-child          5068
 Unmarried          3446
 Wife               1568
 Other-relative      981
Name: relationship, dtype: int64

In [26]:
df["race"].value_counts()

 White                 27816
 Black                  3124
 Asian-Pac-Islander     1039
 Amer-Indian-Eskimo      311
 Other                   271
Name: race, dtype: int64

In [27]:
df["sex"].value_counts()

 Male      21790
 Female    10771
Name: sex, dtype: int64

In [28]:
df["native-country"].value_counts()

 United-States                 29170
 Mexico                          643
 ?                               583
 Philippines                     198
 Germany                         137
 Canada                          121
 Puerto-Rico                     114
 El-Salvador                     106
 India                           100
 Cuba                             95
 England                          90
 Jamaica                          81
 South                            80
 China                            75
 Italy                            73
 Dominican-Republic               70
 Vietnam                          67
 Guatemala                        64
 Japan                            62
 Poland                           60
 Columbia                         59
 Taiwan                           51
 Haiti                            44
 Iran                             43
 Portugal                         37
 Nicaragua                        34
 Peru                             31
 

In [29]:
# replacing "?" with the most frequent value "United-States"
df["native-country"] = df["native-country"].str.replace("?", "United-States")

<ipython-input-29-20de6242a7da>:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df["native-country"] = df["native-country"].str.replace("?", "United-States")


In [30]:
#checking the progress
df["native-country"].value_counts()

 United-States                 29753
 Mexico                          643
 Philippines                     198
 Germany                         137
 Canada                          121
 Puerto-Rico                     114
 El-Salvador                     106
 India                           100
 Cuba                             95
 England                          90
 Jamaica                          81
 South                            80
 China                            75
 Italy                            73
 Dominican-Republic               70
 Vietnam                          67
 Guatemala                        64
 Japan                            62
 Poland                           60
 Columbia                         59
 Taiwan                           51
 Haiti                            44
 Iran                             43
 Portugal                         37
 Nicaragua                        34
 Peru                             31
 Greece                           29
 

In [31]:
df["income"].value_counts()

 <=50K    24720
 >50K      7841
Name: income, dtype: int64

<h4>Observation: </h4>
<p>The columns "relationship", "race", "sex" and "income" match with the attribute information found in https://archive.ics.uci.edu/ml/datasets/Census+Income.</p>

<h3>1.5 Encoding</h3>
<p>Converting categorial features into numerical vectors for machine learning </p>

In [32]:
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder
le = preprocessing.LabelEncoder()
for col in df.columns:
    df[col] = le.fit_transform(df[col])
df.head()

age  workclass  fnlwgt  education  education-num  marital-status  \
0   22          7    2671          9             12               4   
1   33          6    2926          9             12               2   
2   21          4   14086         11              8               0   
3   36          4   15336          1              6               2   
4   11          4   19355          9             12               2   

   occupation  relationship  race  sex  capital-gain  capital-loss  \
0           0             1     4    1            25             0   
1           3             0     4    1             0             0   
2           5             1     4    1             0             0   
3           5             0     2    1             0             0   
4           9             5     2    0             0             0   

   hours-per-week  native-country  income  
0              39              38       0  
1              12              38       0  
2              39              38       0  
3              39              38       0  
4              39               4       0

<h3>1.6 Feature selection</h3>
<p>According to the requirements, 10 columns are needed for machine learning.</p>
<p>Removing features that are redundant, which indicating features that are similar or highly correlated with other features.  </p>

In [33]:
#checking the correlation
df[df.columns[1:]].corr()["income"][:].sort_values(ascending=False)

income            1.000000
capital-gain      0.340016
education-num     0.335154
hours-per-week    0.232393
sex               0.215980
capital-loss      0.162514
education         0.079317
race              0.071846
workclass         0.051604
occupation        0.034625
native-country    0.023058
fnlwgt           -0.010549
marital-status   -0.199307
relationship     -0.250918
Name: income, dtype: float64

<h4>Observation: </h4>
<p>Removing the last 4 columns "native-country", "fnlwgt", "marital-status","relationship" which have the least correlation with the target "income".</p>

In [34]:
#creating a dataset with 10 selected columns for machine learning
df_model = df[["capital-gain","education-num","hours-per-week","sex","capital-loss","education","race","workclass","occupation","income"]]

In [35]:
#checking the new dataframe
df_model.head()

capital-gain  education-num  hours-per-week  sex  capital-loss  education  \
0            25             12              39    1             0          9   
1             0             12              12    1             0          9   
2             0              8              39    1             0         11   
3             0              6              39    1             0          1   
4             0             12              39    0             0          9   

   race  workclass  occupation  income  
0     4          7           0       0  
1     4          6           3       0  
2     4          4           5       0  
3     2          4           5       0  
4     2          4           9       0

In [36]:
df_model.shape

(32561, 10)

<h1>2 Model Development</h1>

<h3>2.1 Maching learning task</h3>
<p>(1)The task is predicting if the income will exceeds 50K or not, which is a binary classification problem.</p>
<p>(2)Since every instance in the dataset consists of a set of feature vectors and its associated label, the machine learning task is a supervised learning.</p>

<h3>2.2 Data Partition</h3>
<h4>(1)hold-out</h4>
<p>Based on the observation from section 1, there are 32561 instances in the dataset. Different splitting ratio of training and test data will not lead to greater variance. </p>
<p>Thus, 80% for training and 20% for testing will be used for data partition.</p>

In [37]:
#getting the feature matrix "X" and the target variable "y"
X = df_model.iloc[:, 0:-1].values
y = df_model.iloc[:, -1].values

In [38]:
# getting Suite: 80% for training and 20% for testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

In [39]:
# printing the training data-features
print(X_train)

[[ 0  6 39 ...  4  6 13]
 [ 0 12 39 ...  4  6  3]
 [ 0  8 49 ...  4  4 11]
 ...
 [ 0 12 39 ...  1  4  6]
 [ 0  8 39 ...  4  7  7]
 [ 0 12 44 ...  4  4  0]]


In [40]:
#getting the size of the training data
print(len(X_train))

26048


In [41]:
# printing the training data-target
print(y_train)

[0 1 1 ... 0 0 1]


In [42]:
#getting the size of the training data
print(len(y_train))

26048


<h4>Observation: </h4>
<p>The traing data has 26048 instances, which is the 80% of the whole dataset.</p>

In [43]:
# printing the test data-features
print(X_test)

[[  0   3  65 ...   4   6   7]
 [  0   6  24 ...   4   4   7]
 [115  12  49 ...   4   4   4]
 ...
 [  0   8  39 ...   4   4  13]
 [  0   6  44 ...   4   4   2]
 [  0  15  59 ...   4   6   9]]


In [44]:
#getting the size of the test data
print(len(X_test))

6513


In [45]:
# printing the test data-target
print(y_test)

[0 0 1 ... 0 0 1]


In [46]:
#getting the size of the test data
print(len(y_test))

6513


<h4>Observation: </h4>
<p>The test data has 6513 instances, which is the 20% of the whole dataset.</p>

<h4>(2)K-fold cross validation</h4>

<P>1. Using K-fold cross validation within the training data, where n(or K) is from 3 to 5.</P>
<p>2. Learning algorithm list</p>
<p>(1) Considerring the learning type of the problem is a binary classification problem, predicting the result to be over 50K or under 50k, three classification algorithms chosen are:</p>
<li>KNN</li>
<li>Logistic Regression</li>
<li>Random Forest</li>
<p>(2) the corresponding hyperparameters are: </p>
<li>KNN: n_neighbors, weights, metric</li>
<li>Logistic Regression: penalty, solver, max_iter, C</li>
<li>Random Forest: n_estimators, max_features</li>
<P>3. Starting with default values or random values for hyperparameters which will be tuned afterwards.</P>
<P>4. Assessing the validation performance for the three algorithms.</p>

<h4><li>Applying n = 3 for K-fold cross-validation</li></h4>

In [47]:
from sklearn.model_selection import StratifiedKFold
skfCV_3 = StratifiedKFold(n_splits=3)

In [48]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=6)

from sklearn.linear_model import LogisticRegression
reg = LogisticRegression(solver='lbfgs', max_iter=400)

from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(n_estimators=100)

In [49]:
score_list_knn=[]
score_list_reg=[]
score_list_rfc=[]

for train_index, test_index in skfCV_3.split(X,y): 
    print("trData_cv3", train_index, "vlData_cv3", train_index)
    X_train_cv3, X_test_cv3 = X[train_index], X[test_index]
    y_train_cv3, y_test_cv3 = y[train_index], y[test_index]    
    knn.fit(X_train_cv3,y_train_cv3) #training KNN on the trDATA
    score_list_knn.append(knn.score(X_test_cv3,y_test_cv3))
    reg.fit(X_train_cv3,y_train_cv3) #training Logistic Regression on the trDATA
    score_list_reg.append(reg.score(X_test_cv3,y_test_cv3))
    rfc.fit(X_train_cv3,y_train_cv3) #training Random Forest on the trDATA
    score_list_rfc.append(rfc.score(X_test_cv3,y_test_cv3))

trData_cv3 [10807 10809 10810 ... 32558 32559 32560] vlData_cv3 [10807 10809 10810 ... 32558 32559 32560]
trData_cv3 [    0     1     2 ... 32558 32559 32560] vlData_cv3 [    0     1     2 ... 32558 32559 32560]
trData_cv3 [    0     1     2 ... 21912 21927 21938] vlData_cv3 [    0     1     2 ... 21912 21927 21938]


In [50]:
avg_P_knn_cv3 = np.mean(score_list_knn,axis=0)
print("KNN'S avarage of P over all folds is:", avg_P_knn_cv3)
avg_P_reg_cv3 = np.mean(score_list_reg,axis=0)
print("Logistic Regression's avarage of P over all folds is:", avg_P_reg_cv3)
avg_P_rfc_cv3 = np.mean(score_list_rfc,axis=0)
print("Random Forest's avarage of P over all folds is:", avg_P_rfc_cv3)

KNN'S avarage of P over all folds is: 0.8231013576391176
Logistic Regression's avarage of P over all folds is: 0.8178802906045863
Random Forest's avarage of P over all folds is: 0.8271245341046983


<h4>Observation: </h4>
<p>Random Forest model has the highest validation performance.</p>

<h4><li>Applying n = 4 for K-fold cross-validation</li></h4>

In [51]:
from sklearn.model_selection import StratifiedKFold
skfCV_4 = StratifiedKFold(n_splits=4)

In [52]:
score_list_knn=[]
score_list_reg=[]
score_list_rfc=[]

for train_index, test_index in skfCV_4.split(X,y): 
    print("trData_cv4", train_index, "vlData_cv4", train_index)
    X_train_cv4, X_test_cv4 = X[train_index], X[test_index]
    y_train_cv4, y_test_cv4 = y[train_index], y[test_index]    
    knn.fit(X_train_cv4,y_train_cv4) #training KNN on the trDATA
    score_list_knn.append(knn.score(X_test_cv4,y_test_cv4))
    reg.fit(X_train_cv4,y_train_cv4) #training Logistic Regression on the trDATA
    score_list_reg.append(reg.score(X_test_cv4,y_test_cv4))
    rfc.fit(X_train_cv4,y_train_cv4) #training Random Forest on the trDATA
    score_list_rfc.append(rfc.score(X_test_cv4,y_test_cv4))

trData_cv4 [ 8123  8126  8127 ... 32558 32559 32560] vlData_cv4 [ 8123  8126  8127 ... 32558 32559 32560]
trData_cv4 [    0     1     2 ... 32558 32559 32560] vlData_cv4 [    0     1     2 ... 32558 32559 32560]
trData_cv4 [    0     1     2 ... 32558 32559 32560] vlData_cv4 [    0     1     2 ... 32558 32559 32560]
trData_cv4 [    0     1     2 ... 24510 24511 24514] vlData_cv4 [    0     1     2 ... 24510 24511 24514]


In [53]:
avg_P_knn_cv4 = np.mean(score_list_knn,axis=0)
print("KNN'S avarage of P over all folds is:", avg_P_knn_cv4)
avg_P_reg_cv4 = np.mean(score_list_reg,axis=0)
print("Logistic Regression's avarage of P over all folds is:", avg_P_reg_cv4)
avg_P_rfc_cv4 = np.mean(score_list_rfc,axis=0)
print("Random Forest's avarage of P over all folds is:", avg_P_rfc_cv4)

KNN'S avarage of P over all folds is: 0.8241455118282289
Logistic Regression's avarage of P over all folds is: 0.8182488455166873
Random Forest's avarage of P over all folds is: 0.8259268801078776


<h4>Observation: </h4>
<p>Random Forest model still has the highest validation performance.</p>

<h4><li>Applying n = 5 for K-fold cross-validation</li></h4>

In [54]:
from sklearn.model_selection import StratifiedKFold
skfCV_5 = StratifiedKFold(n_splits=5)

In [55]:
score_list_knn=[]
score_list_reg=[]
score_list_rfc=[]

for train_index, test_index in skfCV_5.split(X,y): 
    print("trData_cv5", train_index, "vlData_cv5", train_index)
    X_train_cv5, X_test_cv5 = X[train_index], X[test_index]
    y_train_cv5, y_test_cv5 = y[train_index], y[test_index]    
    knn.fit(X_train_cv5,y_train_cv5) #training KNN on the trDATA
    score_list_knn.append(knn.score(X_test_cv5,y_test_cv5))
    reg.fit(X_train_cv5,y_train_cv5) #training Logistic Regression on the trDATA
    score_list_reg.append(reg.score(X_test_cv5,y_test_cv5))
    rfc.fit(X_train_cv5,y_train_cv5) #training Random Forest on the trDATA
    score_list_rfc.append(rfc.score(X_test_cv5,y_test_cv5))

trData_cv5 [ 6499  6500  6512 ... 32558 32559 32560] vlData_cv5 [ 6499  6500  6512 ... 32558 32559 32560]
trData_cv5 [    0     1     2 ... 32558 32559 32560] vlData_cv5 [    0     1     2 ... 32558 32559 32560]
trData_cv5 [    0     1     2 ... 32558 32559 32560] vlData_cv5 [    0     1     2 ... 32558 32559 32560]
trData_cv5 [    0     1     2 ... 32558 32559 32560] vlData_cv5 [    0     1     2 ... 32558 32559 32560]
trData_cv5 [    0     1     2 ... 26161 26168 26170] vlData_cv5 [    0     1     2 ... 26161 26168 26170]


In [56]:
avg_P_knn_cv5 = np.mean(score_list_knn,axis=0)
print("KNN'S avarage of P over all folds is:", avg_P_knn_cv5)
avg_P_reg_cv5 = np.mean(score_list_reg,axis=0)
print("Logistic Regression's avarage of P over all folds is:", avg_P_reg_cv5)
avg_P_rfc_cv5 = np.mean(score_list_rfc,axis=0)
print("Random Forest's avarage of P over all folds is:", avg_P_rfc_cv5)

KNN'S avarage of P over all folds is: 0.8235619151038314
Logistic Regression's avarage of P over all folds is: 0.8177881526683922
Random Forest's avarage of P over all folds is: 0.8252204719270587


<h4>Observation: </h4>
<p>Random Forest model still has the highest validation performance.</p>

<h4>Summary</h4>
<p>1. Random Forest model has the highest validation perfomance compared with KNN and logistic regression.</p>
<p>2. Random Forest model's performance is slightly better when n=3 (score=0.8270324106608454)than n=4(score=0.8261725237649571). Thus, use K-fold cross-validation(n=3) will be used for hyperparameter tuning.</p>

<h3>2.3 Hyperparameter tuning</h3>

<h4>2.3.1 KNN's hyperparameter tuning</h4>

<p><strong>For KNN:</strong></p>
<li>n_neighbors: [6 to 10]</li>
<li>weights:["uniform", "distance"]</li>
<li>metric:["euclidean", "manhattan", "minkowski"]</li>

In [57]:
from sklearn.model_selection import GridSearchCV
knn_hyper = KNeighborsClassifier()
n_neighbors = range(6,10)
weights = ["uniform", "distance"]
metric = ["euclidean", "manhattan", "minkowski"]
# define grid search
grid = dict(n_neighbors=n_neighbors,weights=weights,metric=metric)

grid_search = GridSearchCV(estimator=knn_hyper, param_grid=grid, n_jobs=-1, cv=skfCV_3, scoring='accuracy',error_score=0)
grid_result = grid_search.fit(X, y)
# summarize results
print("Best hyperparameter setting for KNN is: %f using %s" % (grid_result.best_score_, grid_result.best_params_))

Best hyperparameter setting for KNN is: 0.829919 using {'metric': 'manhattan', 'n_neighbors': 8, 'weights': 'uniform'}


<h4>2.3.2 Logistic Regression's hyperparameter tuning</h4>

<p><strong>For Logistic Regression:</strong></p>
<li>penalty: ["l1", "l2", "elasticnet", "none"]</li>
<li>solver:["newton-cg", "lbfgs", "liblinear"]</li>
<li>max_iter: [100, 200, 300, 400,500]</li>
<li>C: [100, 10, 1.0, 0.1, 0.01]</li>

In [85]:
from sklearn.model_selection import GridSearchCV
reg_hyper = LogisticRegression()
penalty = ["l1", "l2", "elasticnet", "none"]
solver = ["newton-cg", "liblinear"]
max_iter = [100, 200, 300, 400,500]
C = [100, 10, 1.0, 0.1, 0.01]

# define grid search
grid = dict(penalty=penalty, solver=solver, max_iter=max_iter, C=C )

grid_search = GridSearchCV(estimator=reg_hyper, param_grid=grid, n_jobs=-1, cv=skfCV_3, scoring='accuracy',error_score=0)
grid_result = grid_search.fit(X, y)
# summarize results
print("Best hyperparameter setting for LogisticRegression is: %f using %s" % (grid_result.best_score_, grid_result.best_params_))

Best hyperparameter setting for LogisticRegression is: 0.818034 using {'C': 0.1, 'max_iter': 300, 'penalty': 'l1', 'solver': 'liblinear'}


<h4>2.3.3 Random Forest's hyperparameter tuning</h4>

<p><strong>For Random Forest:</strong></p>
<li>n_estimators: [100, 200, 500, 1000]</li>
<li>max_features: ["sqrt", "log2", None]</li>

In [59]:
from sklearn.model_selection import GridSearchCV
rfc_hyper = RandomForestClassifier()
n_estimators = [100, 200, 500, 1000]
max_features = ["sqrt", "log2", None]

# define grid search
grid = dict(n_estimators=n_estimators, max_features=max_features)

grid_search = GridSearchCV(estimator=rfc_hyper, param_grid=grid, n_jobs=-1, cv=skfCV_3, scoring='accuracy',error_score=0)
grid_result = grid_search.fit(X, y)
# summarize results
print("Best hyperparameter setting for Random Forest is: %f using %s" % (grid_result.best_score_, grid_result.best_params_))

Best hyperparameter setting for Random Forest is: 0.827462 using {'max_features': 'sqrt', 'n_estimators': 500}


<h4>Summary</h4>
<p>The hyperparameter values of the three algorithms after tunning process are:</p> 
<li>KNN: n_neighbors=8, weights="uniform", metric="manhattan"</li>
<li>Logistic Regression: penalty="l1", solver="liblinear", max_iter=300, C=0.1</li>
<li>Random Forest: n_estimators=500, max_features="sqrt"</li>

<h3>2.4  Reassessing the validation performance for the three algorithms with tuned hyperparameter</h3>
<P>1. Define an n-fold cross validation within the training data, where n is from 3 to 5.</P>
<P>2. Tuned hyperparameter used for the three algorithms are:</P>
<li>KNN: n_neighbors=8, weights="uniform", metric="manhattan"</li>
<li>Logistic Regression: penalty="l1", solver="liblinear", max_iter=300, C=0.1</li>
<li>Random Forest: n_estimators=500, max_features="sqrt"</li>

<h4>(1)Applying n = 3 for K-fold cross-validation</h4>

In [60]:
from sklearn.model_selection import StratifiedKFold
skfCV_3_reass = StratifiedKFold(n_splits=3)

In [86]:
from sklearn.neighbors import KNeighborsClassifier
knn_tunned = KNeighborsClassifier(n_neighbors=8, weights="uniform", metric="manhattan")

from sklearn.linear_model import LogisticRegression
reg_tunned = LogisticRegression(penalty="l1", solver="liblinear", max_iter=300, C=0.1)

from sklearn.ensemble import RandomForestClassifier
rfc_tunned = RandomForestClassifier(n_estimators=500, max_features="sqrt")

In [87]:
score_list_knn_reass=[]
score_list_reg_reass=[]
score_list_rfc_reass=[]

for train_index, test_index in skfCV_3_reass.split(X,y): 
    print("trData_cv3_reass", train_index, "vlData_cv3_reass", train_index)
    X_train_cv3, X_test_cv3 = X[train_index], X[test_index]
    y_train_cv3, y_test_cv3 = y[train_index], y[test_index]    
    knn_tunned.fit(X_train_cv3,y_train_cv3) #training KNN on the trDATA
    score_list_knn_reass.append(knn_tunned.score(X_test_cv3,y_test_cv3))
    reg_tunned.fit(X_train_cv3,y_train_cv3) #training Logistic Regression on the trDATA
    score_list_reg_reass.append(reg_tunned.score(X_test_cv3,y_test_cv3))
    rfc_tunned.fit(X_train_cv3,y_train_cv3) #training Random Forest on the trDATA
    score_list_rfc_reass.append(rfc_tunned.score(X_test_cv3,y_test_cv3))

trData_cv3_reass [10807 10809 10810 ... 32558 32559 32560] vlData_cv3_reass [10807 10809 10810 ... 32558 32559 32560]
trData_cv3_reass [    0     1     2 ... 32558 32559 32560] vlData_cv3_reass [    0     1     2 ... 32558 32559 32560]
trData_cv3_reass [    0     1     2 ... 21912 21927 21938] vlData_cv3_reass [    0     1     2 ... 21912 21927 21938]


In [88]:
avg_P_knn_cv3_reass = np.mean(score_list_knn_reass,axis=0)
print("KNN'S avarage of P over all folds is:", avg_P_knn_cv3_reass)
avg_P_reg_cv3_reass = np.mean(score_list_reg_reass,axis=0)
print("Logistic Regression's avarage of P over all folds is:", avg_P_reg_cv3_reass)
avg_P_rfc_cv3_reass = np.mean(score_list_rfc_reass,axis=0)
print("Random Forest's avarage of P over all folds is:", avg_P_rfc_cv3_reass)

KNN'S avarage of P over all folds is: 0.8299193611995829
Logistic Regression's avarage of P over all folds is: 0.8179417118932051
Random Forest's avarage of P over all folds is: 0.8276773738070252


<h4>Observation: </h4>
<p>After tuning, KNN has the highest validation performance with score 0.8299193611995829.</p>

<h4>(2)Applying n = 4 for K-fold cross-validation</h4>

In [64]:
from sklearn.model_selection import StratifiedKFold
skfCV_4_reass = StratifiedKFold(n_splits=4)

In [89]:
score_list_knn_reass=[]
score_list_reg_reass=[]
score_list_rfc_reass=[]

for train_index, test_index in skfCV_4_reass.split(X,y): 
    print("trData_cv4_reass", train_index, "vlData_cv4_reass", train_index)
    X_train_cv4, X_test_cv4 = X[train_index], X[test_index]
    y_train_cv4, y_test_cv4 = y[train_index], y[test_index]    
    knn_tunned.fit(X_train_cv4,y_train_cv4) #training KNN on the trDATA
    score_list_knn_reass.append(knn_tunned.score(X_test_cv4,y_test_cv4))
    reg_tunned.fit(X_train_cv4,y_train_cv4) #training Logistic Regression on the trDATA
    score_list_reg_reass.append(reg_tunned.score(X_test_cv4,y_test_cv4))
    rfc_tunned.fit(X_train_cv4,y_train_cv4) #training Random Forest on the trDATA
    score_list_rfc_reass.append(rfc_tunned.score(X_test_cv4,y_test_cv4))

trData_cv4_reass [ 8123  8126  8127 ... 32558 32559 32560] vlData_cv4_reass [ 8123  8126  8127 ... 32558 32559 32560]
trData_cv4_reass [    0     1     2 ... 32558 32559 32560] vlData_cv4_reass [    0     1     2 ... 32558 32559 32560]
trData_cv4_reass [    0     1     2 ... 32558 32559 32560] vlData_cv4_reass [    0     1     2 ... 32558 32559 32560]
trData_cv4_reass [    0     1     2 ... 24510 24511 24514] vlData_cv4_reass [    0     1     2 ... 24510 24511 24514]


In [90]:
avg_P_knn_cv4_reass = np.mean(score_list_knn_reass,axis=0)
print("KNN'S avarage of P over all folds is:", avg_P_knn_cv4_reass)
avg_P_reg_cv4_reass = np.mean(score_list_reg_reass,axis=0)
print("Logistic Regression's avarage of P over all folds is:", avg_P_reg_cv4_reass)
avg_P_rfc_cv4_reass = np.mean(score_list_rfc_reass,axis=0)
print("Random Forest's avarage of P over all folds is:", avg_P_rfc_cv4_reass)

KNN'S avarage of P over all folds is: 0.8290901085505558
Logistic Regression's avarage of P over all folds is: 0.8178802989207117
Random Forest's avarage of P over all folds is: 0.8251590592949148


<h4>Observation: </h4>
<p>With tuned hyperperameter, KNN has the highest validation performance with score 0.8290901085505558.</p>

<h4>(3)Applying n = 5 for K-fold cross-validation</h4>

In [67]:
from sklearn.model_selection import StratifiedKFold
skfCV_5_reass = StratifiedKFold(n_splits=5)

In [91]:
score_list_knn_reass=[]
score_list_reg_reass=[]
score_list_rfc_reass=[]

for train_index, test_index in skfCV_5_reass.split(X,y): 
    print("trData_cv5_reass", train_index, "vlData_cv5_reass", train_index)
    X_train_cv5, X_test_cv5 = X[train_index], X[test_index]
    y_train_cv5, y_test_cv5 = y[train_index], y[test_index]    
    knn_tunned.fit(X_train_cv5,y_train_cv5) #training KNN on the trDATA
    score_list_knn_reass.append(knn_tunned.score(X_test_cv5,y_test_cv5))
    reg_tunned.fit(X_train_cv5,y_train_cv5) #training Logistic Regression on the trDATA
    score_list_reg_reass.append(reg_tunned.score(X_test_cv5,y_test_cv5))
    rfc_tunned.fit(X_train_cv5,y_train_cv5) #training Random Forest on the trDATA
    score_list_rfc_reass.append(rfc_tunned.score(X_test_cv5,y_test_cv5))

trData_cv5_reass [ 6499  6500  6512 ... 32558 32559 32560] vlData_cv5_reass [ 6499  6500  6512 ... 32558 32559 32560]
trData_cv5_reass [    0     1     2 ... 32558 32559 32560] vlData_cv5_reass [    0     1     2 ... 32558 32559 32560]
trData_cv5_reass [    0     1     2 ... 32558 32559 32560] vlData_cv5_reass [    0     1     2 ... 32558 32559 32560]
trData_cv5_reass [    0     1     2 ... 32558 32559 32560] vlData_cv5_reass [    0     1     2 ... 32558 32559 32560]
trData_cv5_reass [    0     1     2 ... 26161 26168 26170] vlData_cv5_reass [    0     1     2 ... 26161 26168 26170]


In [92]:
avg_P_knn_cv5_reass = np.mean(score_list_knn_reass,axis=0)
print("KNN'S avarage of P over all folds is:", avg_P_knn_cv5_reass)
avg_P_reg_cv5_reass = np.mean(score_list_reg_reass,axis=0)
print("Logistic Regression's avarage of P over all folds is:", avg_P_reg_cv5_reass)
avg_P_rfc_cv5_reass = np.mean(score_list_rfc_reass,axis=0)
print("Random Forest's avarage of P over all folds is:", avg_P_rfc_cv5_reass)

KNN'S avarage of P over all folds is: 0.8299500601895812
Logistic Regression's avarage of P over all folds is: 0.8176038916308377
Random Forest's avarage of P over all folds is: 0.8251591458926789


<h4>Observation: </h4>
<p>With tuned hyperperameter, KNN has the highest validation performance with score 0.8299500601895812.</p>

<h4>Summary</h4>
<p>With tuned hyperperameter, KNN has the highest validation performance compared with Logistic Regression and Random Forest with score 0.82995. And K-fold cross validation with n=5 has the best result.</p>

<h1>3 Performance Assessment</h1>

<h3>3.1 Data Prediction</h3>
<p>1. Apply KNN on the test data to get the prediction.</p>
<p>2. The trained data and test data used are the ones with n=5 for K-fold cross validation.</p>

In [70]:
#using "knn_tunned" for KNN model with tuned hyperperameters for prediction on test data prediction
knn_trained = knn_tunned.fit(X_train_cv5,y_train_cv5)
y_test_pred = knn_trained.predict(X_test_cv5)

In [71]:
#Add a column in the test data name "predicted" which holds the predictions for each row
df_for_X_test = pd.DataFrame(X_test_cv5)
df_for_y_test = pd.DataFrame(y_test_cv5)
df_for_y_test_pred = pd.DataFrame(y_test_pred)

df_testdata = pd.concat([df_for_X_test, df_for_y_test], axis=1)
df_testdata["predictied"] = df_for_y_test_pred

In [72]:
#Printing out the combined test dataset
df_testdata 

0   1   2  3  4   5  6  7   8  0  predictied
0       0  10  39  1  0   8  4  4  11  0           0
1       0   8  33  0  0  11  4  4   0  0           0
2       0  13  39  1  0  12  1  4  11  0           0
3       0   8  39  1  0  11  4  4   7  0           0
4       0   6  29  1  0   1  4  4  11  0           0
...   ...  ..  .. .. ..  .. .. ..  .. ..         ...
6507    0  11  37  0  0   7  4  4  12  0           0
6508    0   8  39  1  0  11  4  4   6  1           0
6509    0   8  39  0  0  11  4  4   0  0           0
6510    0   8  19  1  0  11  4  4   0  0           0
6511  108   8  39  0  0  11  4  5   3  1           1

[6512 rows x 11 columns]

<h3>3.2 Performance Evaluation</h3>
<p>Evaluating the performances of the model on the training and test dataset with the following metrics:</p>
<li>Confusion matrix</li>
<li>Classification accuracy</li>
<li>Precision</li>
<li>Recall</li>
<li>F1 score</li>

<h4>(1)Confusion matrix</h4>

In [73]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test_cv5, y_test_pred) 

array([[4705,  239],
       [ 852,  716]], dtype=int64)

<h4>(2)Classification accuracy</h4>

In [74]:
from sklearn.metrics import accuracy_score 
acc_score = accuracy_score(y_test_cv5, y_test_pred) 

print("The accuracy score is:", acc_score) 

The accuracy score is: 0.832463144963145


<h4>(3)Precision</h4>

In [75]:
from sklearn.metrics import precision_score
pre_score = precision_score(y_test_cv5, y_test_pred, average="macro") 
print("The precision score is:", pre_score) 

The precision score is: 0.7982090415654234


<h4>(4)Recall</h4>

In [76]:
from sklearn.metrics import recall_score
rec_score = recall_score(y_test_cv5, y_test_pred, average="macro")
print("The recall score is:", rec_score) 

The recall score is: 0.7041456145565022


<h4>(5)F1 score</h4>

In [77]:
from sklearn.metrics import f1_score
f1_score = f1_score(y_test_cv5, y_test_pred, average="macro") 
print("The F1 score is:", f1_score) 

The F1 score is: 0.7318417063350477


<h4>(6)Classification report</h4>

In [78]:
from sklearn.metrics import classification_report
print("classification report")
print(classification_report(y_test_cv5, y_test_pred))

classification report
              precision    recall  f1-score   support

           0       0.85      0.95      0.90      4944
           1       0.75      0.46      0.57      1568

    accuracy                           0.83      6512
   macro avg       0.80      0.70      0.73      6512
weighted avg       0.82      0.83      0.82      6512



In [79]:
#decode the target value from the binary values
le.classes_

array([' <=50K', ' >50K'], dtype=object)

In [80]:
le.inverse_transform([0,1])

array([' <=50K', ' >50K'], dtype=object)

In [81]:
dict(zip(le.inverse_transform([0,1]),[0,1]))

{' <=50K': 0, ' >50K': 1}

<h4>Summary:</h4>
<li>With the classfier KNN on the test data, all scores are above 0.70, indicating a good performance on the classfication prediction.</li>
<li>The "support"values in the classification report the two classes "0"(<=50K) and "1"(>50K) shows that the dataset is not a well balanced one.</li>
<li>Scores for class "0"(<=50K) are higher than class "1"(>50K) indicating the algoritem KNN is better in finding class "0" over class"1".</li>    

<h1>4 Conclusion</h1>
<p>(1)The "Census Income Data Set" used in the project is used for predicting if the income will exceeds 50K or not.</p>
<p>(2)This is a classification problem. Supervised learning is used to solve the problem.</p>
<p>(3)The dataset has 32561 rows and 15 columns. Among the 14 columns for feature selection, 10 features with the highest correlation scores are chosen for data modelling.</p>
<p>(4)For data partition, K-fold cross validation is chosen over the holdout.</p>
<p>(5)Three classifers used in the project are: KNeighbors Classifier(KNN), Logistic Regression and RandomForest Classifier.</p>
<p>(6)K-fold cross validation is applied for hyperperameter tuning.</p>
<p>(7)The classfier KNN shows the best performance, which is used for test data prediction and perform well in the prediction.</p>

In [82]:
#export the file into a binary file
import pickle
with open("model","wb") as f:
    pickle.dump(knn_tunned,f)